### PR TITLE
editor/code: Enable `--noUncheckedIndexedAccess` & `--noPropertyAccessFromIndexSignature` ts option

### PR DESCRIPTION
**Pull request by tetsuharuohzeki**
This enables typescript's these option:

- [`--noUncheckedIndexedAccess`](https://www.typescriptlang.org/tsconfig#noUncheckedIndexedAccess)
    - This checks whether indexed access is not `null` (or `undefined`) strictly just as like checking by `std::option::Option::unwrap()`.
- [`--noPropertyAccessFromIndexSignature`](https://www.typescriptlang.org/tsconfig#noPropertyAccessFromIndexSignature)
    - This disallows `bar.foo` access if the `bar` type is `{ [key: string]: someType; }`.

----

Additionally, to enable `--noUncheckedIndexedAccess` easily, this pull request introduces [option-t](https://www.npmjs.com/package/option-t) as a dependency instead of defining a function in this repository like `unwrapUndefinable()` .

I'll remove it and define them byself if our dependency management policy is that to avoid to add a new package as possible.

### CODE REVIEW

File: editors/code/tsconfig.json [hunk 1]
```diff
@@ -3,6 +3,7 @@
     "compilerOptions": {
         "esModuleInterop": false,
         "module": "commonjs",
+        "moduleResolution": "Node16",
```
Comment (lnicola) [inline editors/code/tsconfig.json:6]: Should this be `node16` as in the docs?

**Comment by lnicola** [comment]
LGTM except for the `noUncheckedIndexedAccess` commit, which feels really heavy-handed. But I guess I'm not opposed to that either.

r? @Veykril

File: editors/code/tsconfig.json (hunk 1 above)
Comment (tetsuharuohzeki) [reply editors/code/tsconfig.json:6]: Ah, there is some reasons:

1. Future enhancement. This mode (`node16`) is more natural behavior with the current Node.js' package resolution than `node` (the current default value). And [package.json's `exports`](https://nodejs.org/api/packages.html#exports-sugar) would be a wide spreading. I think it's better to enable as early as possible.
2. Compatibility with `option-t` that depends on `exports` field too.

Comment (lnicola) [reply editors/code/tsconfig.json:6]: Sorry -- I meant `node16` instead of `Node16`, with a lower-case `n`.

Comment (tetsuharuohzeki) [reply editors/code/tsconfig.json:6]: (Mysteriously), it works in either. TypeScript compiler use lower-case but vscode's intellisense uses CamelCase.
I'll address it to lower-case.

Comment (lnicola) [reply editors/code/tsconfig.json]: I guess either is fine then as long as nothing complains.

Comment (tetsuharuohzeki) [reply editors/code/tsconfig.json]: I pushed the addressed commit: https://github.com/rust-lang/rust-analyzer/pull/15160/commits/f780810bce11bd219782e824709c36846da3b192

**Comment by lowr** [comment]
I'm in favor of `noUncheckedIndexedAccess`, but I feel we can do better by considering the context instead of unconditionally "unwrapping" nullables. Unwrapping after checking array length/property existence/etc. seems redundant. For example, imo

```ts
// simplified for brevity
if (folders.length === 1) {
    const folder = unwrapUndefinable(folders[0]);
    return folder.uri.fsPath;
} else if (folders.length > 1) {
    const folder = unwrapUndefinable(folders[0]);
    // comments for justification
    return folder.uri.fsPath;
} else {
    return "";
}
```

should be 

```ts
// comments for justification
const folder = folders[0];
return folder != null ? folder.uri.fsPath : "";
```

Also, I'd rather not pull in another dependency for a utility function or two, but I'll leave the decision to others.

**Comment by Veykril** [comment]
> Also, I'd rather not pull in another dependency for a utility function or two, but I'll leave the decision to others.

I agree here, we only use a single function from that library if I see this right, so if we can't get rid of the call locations for that, can we just hoist that function out into the local source here and drop the dependency?

**Comment by tetsuharuohzeki** [comment]
@Veykril 

> I agree here, we only use a single function from that library if I see this right, so if we can't get rid of the call locations for that, can we just hoist that function out into the local source here and drop the dependency?

Sure. I addressed them and rebased my changeset again.

**Review by Veykril — COMMENTED** [review]
One small question, but otherwise this should be good for now. Ideally we should get rid of the unwraps where we can, but most of them seem to be due to ts not being smart enough (like array length checks into indexing still resulting in undefined types)

File: editors/code/src/commands.ts [hunk 2]
```diff
@@ -129,7 +130,8 @@ export function matchingBrace(ctx: CtxInit): Cmd {
             ),
         });
         editor.selections = editor.selections.map((sel, idx) => {
-            const active = client.protocol2CodeConverter.asPosition(response[idx]);
+            const mayBeActive = client.protocol2CodeConverter.asPosition(response[idx]);
```
[the remark this answers is not in the document]
Comment (tetsuharuohzeki) [reply editors/code/src/commands.ts:133]: @Veykril 

At here, `response[idx]` should be a `ls.Position | undefined`. Then TypeScript selects `asPosition(value: ls.Position | undefined | null): code.Position | undefined;` overload than `asPosition(value: ls.Position): code.Position;`.
https://github.com/microsoft/vscode-languageserver-node/blob/2041784436fed53f4e77267a49396bca22a7aacf/client/src/common/protocolConverter.ts#L34-L36

[the remark this answers is not in the document]
Comment (tetsuharuohzeki) [reply editors/code/src/commands.ts:133]: I'll update here to:

```diff
         editor.selections = editor.selections.map((sel, idx) => {
-            const mayBeActive = client.protocol2CodeConverter.asPosition(response[idx]);
-            const active = unwrapUndefinable(mayBeActive);
+            const position = unwrapUndefinable(response[idx]);
+            const active = client.protocol2CodeConverter.asPosition(position);
             const anchor = sel.isEmpty ? active : sel.anchor;
             return new vscode.Selection(anchor, active);
         });
```

[the remark this answers is not in the document]
Comment (tetsuharuohzeki) [reply editors/code/src/commands.ts]: addressed: https://github.com/rust-lang/rust-analyzer/compare/edf14df49121e3d0d8d06bd9b650d4d74918a69e..f70845305f54181e1681828c7935305c66e743bb

[the remark this answers is not in the document]
Comment (Veykril) [reply editors/code/src/commands.ts]: Oh i missed the indexing into the array, that explains it sorry. Typescript inference is weird, given it tells me that indexing results in a non-undefined type ...
Nevertheless this change is better, thanks

**Comment by Veykril** [comment]
Thanks
@bors r+

**Pull request merged**